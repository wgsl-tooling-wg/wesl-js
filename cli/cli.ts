import { ModuleRegistry, normalize } from "@wesl/linker";
import { createTwoFilesPatch } from "diff";
import { TypeRefElem } from "../linker/AbstractElems.ts";
import yargs from "yargs";

// TODO: Check if these are correct, and figure out why the types are broken
type CliArgs = {
  define?: (string | number)[];
  baseDir?: string;
  separately: boolean;
  details: boolean;
  diff: boolean;
  emit: boolean;
  files: string[];
};

export async function cli(rawArgs: string[]): Promise<void> {
  const argv = parseArgs(rawArgs);
  const files = argv.files as string[];
  if (argv.separately) await linkSeparately(argv, files);
  else await linkNormally(argv, files);
}

function parseArgs(args: string[]): CliArgs {
  return yargs(args)
    .command(
      "$0 <files...>",
      "root wgsl file followed by any library wgsl files",
    )
    .option("define", {
      type: "array",
      describe: "definitions for preprocessor and linking",
    })
    .option("baseDir", {
      requiresArg: true,
      type: "string",
      describe: "rm common prefix from file paths",
    })
    .option("separately", {
      type: "boolean",
      default: false,
      hidden: true,
      describe: "link each file separately (for parser testing)",
    })
    .option("details", {
      type: "boolean",
      default: false,
      hidden: true,
      describe: "show details about parsed files",
    })
    .option("diff", {
      type: "boolean",
      default: false,
      hidden: true,
      describe: "show comparison with src file",
    })
    .option("emit", {
      type: "boolean",
      default: true,
      hidden: true,
      describe: "emit linked result",
    })
    .help()
    .parseSync();
}

async function linkNormally(argv: CliArgs, paths: string[]): Promise<void> {
  const basedPaths = paths.map((path) => ({
    path: path,
    basedPath: normalize(rmBaseDirPrefix(argv.baseDir, path)),
  }));
  const wgsl: Record<string, string> = {};
  for (const { path, basedPath } of basedPaths) {
    wgsl[basedPath] = await Deno.readTextFile(path);
  }
  const registry = new ModuleRegistry({ wgsl });
  const srcPath = basedPaths[0].basedPath;
  const srcText = wgsl[srcPath];
  doLink(argv, srcPath, registry, srcText);
}

async function linkSeparately(argv: CliArgs, paths: string[]): Promise<void> {
  for (const path of paths) {
    const srcText = await Deno.readTextFile(path);
    const basedPath = normalize(rmBaseDirPrefix(argv.baseDir, path));
    const registry = new ModuleRegistry({ wgsl: { [basedPath]: srcText } });
    doLink(argv, basedPath, registry, srcText);
  }
}

function doLink(
  argv: CliArgs,
  srcPath: string,
  registry: ModuleRegistry,
  origWgsl: string,
): void {
  const asRelative = "./" + srcPath;
  const linked = registry.link(asRelative, externalDefines(argv));
  if (argv.emit) console.log(linked);
  if (argv.diff) printDiff(srcPath, origWgsl, linked);
  if (argv.details) printDetails(argv, srcPath, registry);
}

function externalDefines(argv: CliArgs): Record<string, string> {
  if (!argv.define) return {};
  const pairs = argv.define.map((d) => d.toString().split("="));

  const badPair = pairs.find((p) => p.length !== 2);
  if (badPair) {
    console.error("invalid define", badPair);
    return {};
  }

  const withParsedValues = pairs.map(([k, v]) => [k, parseDefineValue(v)]);
  return Object.fromEntries(withParsedValues);
}

function parseDefineValue(value: string): string | number | boolean {
  const v = value.toLowerCase();
  if (v === "true") return true;
  if (v === "false") return false;
  if (value === "NaN") return NaN;
  const n = Number.parseFloat(value);
  if (!Number.isNaN(n)) return n;
  return value;
}

function printDiff(modulePath: string, src: string, linked: string): void {
  if (src !== linked) {
    const patch = createTwoFilesPatch(modulePath, "linked", src, linked);
    console.log(patch);
  } else {
    console.log(`${modulePath}: linked version matches original source`);
  }
}

function printDetails(
  argv: CliArgs,
  modulePath: string,
  registry: ModuleRegistry,
): void {
  console.log(modulePath, ":");
  const parsed = registry.parsed(externalDefines(argv));
  const m = parsed.findTextModule(modulePath)!;
  m.fns.forEach((f) => {
    console.log(`  fn ${f.name}`);
    const calls = f.calls.map((c) => c.name).join("  ");
    console.log(`    calls: ${calls}`);
    printTypeRefs(f);
  });
  m.vars.forEach((v) => {
    console.log(`  var ${v.name}`);
    printTypeRefs(v);
  });
  m.structs.forEach((s) => {
    console.log(`  struct ${s.name}`);
    const members = (s.members ?? []).map((m) => m.name).join("  ");
    console.log(`    members: ${members}`);
    s.members.map((m) => printTypeRefs(m));
  });
  console.log();
}

function printTypeRefs(hasTypeRefs: { typeRefs: TypeRefElem[] }): void {
  const typeRefs = hasTypeRefs.typeRefs.map((t) => t.name).join("  ");
  console.log(`    typeRefs: ${typeRefs}`);
}

function rmBaseDirPrefix(baseDir: string | undefined, path: string): string {
  if (baseDir) {
    const found = path.indexOf(baseDir);
    if (found !== -1) {
      return path.slice(found + baseDir.length);
    }
  }
  return path;
}
