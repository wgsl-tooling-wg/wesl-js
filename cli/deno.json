{
  "version": "0.1.0",
  "tasks": {
    "dev": "deno run --watch main.ts"
  },
  "imports": {
    "yargs": "https://deno.land/x/yargs@v17.7.2-deno/deno.ts",
    "diff": "npm:diff"
  }
}