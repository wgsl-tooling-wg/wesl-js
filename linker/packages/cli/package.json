{
  "name": "wgsl-link",
  "version": "0.5.0-d1",
  "type": "module",
  "scripts": {
    "build": "esbuild --bundle --platform=node --format=esm --outfile=bin/wgsl-link src/main.ts",
    "format": "prettier . --write",
    "lint": "eslint src",
    "organize": "organize-imports-cli tsconfig.json",
    "prepublishOnly": "run-s build",
    "test": "vitest",
    "test:once": "vitest run"
  },
  "repository": "https://github.com/wgsl-tooling-wg/wgsl-linker/tree/main/packages/cli",
  "homepage": "https://github.com/wgsl-tooling-wg/wgsl-linker/tree/main/packages/cli#readme",
  "files": [
    "bin"
  ],
  "bin": "bin/wgsl-link",
  "dependencies": {
    "@types/diff": "^5.0.9",
    "diff": "^5.2.0",
    "wgsl-linker": "workspace:*"
  }
}
