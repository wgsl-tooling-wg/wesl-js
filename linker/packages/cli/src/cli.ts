import { createTwoFilesPatch } from "diff";
import fs from "fs";
import { enableTracing } from "mini-parse";
import { astToString, link, normalize, scopeToString } from "wgsl-linker";
import yargs from "yargs";
import { log} from "mini-parse";
import {
  parsedRegistry,
  parseIntoRegistry,
} from "../../linker/src/ParsedRegistry.js";

type CliArgs = ReturnType<typeof parseArgs>;
let argv: CliArgs;

export async function cli(rawArgs: string[]): Promise<void> {
  enableTracing(); // so we get more debug info
  argv = parseArgs(rawArgs);
  const files = argv.files as string[];
  linkNormally(files);
}

function parseArgs(args: string[]) {
  return yargs(args)
    .command(
      "$0 <files...>",
      "root wgsl file followed by any library wgsl files",
    )
    .option("define", {
      type: "array",
      describe: "definitions for preprocessor and linking",
    })
    .option("baseDir", {
      requiresArg: true,
      type: "string",
      describe: "rm common prefix from file paths",
    })
    .option("details", {
      type: "boolean",
      default: false,
      hidden: true,
      describe: "show details about parsed files",
    })
    .option("diff", {
      type: "boolean",
      default: false,
      hidden: true,
      describe: "show comparison with src file",
    })
    .option("emit", {
      type: "boolean",
      default: true,
      hidden: true,
      describe: "emit linked result",
    })
    .help()
    .parseSync();
}

function linkNormally(paths: string[]): void {
  const pathAndTexts = paths.map(f => {
    const text = fs.readFileSync(f, { encoding: "utf8" });
    const basedPath = "./" + normalize(rmBaseDirPrefix(f));
    return [basedPath, text];
  });
  const [rootPath] = pathAndTexts[0];
  const weslFiles = Object.fromEntries(pathAndTexts);

  // TODO conditions
  // TODO external defines
  if (argv.emit) {
    const linked = link(weslFiles, rootPath);
    if (argv.emit) log(linked.dest);
  }
  if (argv.details) {
    const registry = parsedRegistry();
    try {
      parseIntoRegistry(weslFiles, registry, "package");
    } catch (e) {
      console.error(e);
    }
    Object.entries(registry.modules).forEach(([modulePath, ast]) => {
      log(`---\n${modulePath}`);
      log(`\n->ast`);
      log(astToString(ast.moduleElem));
      log(`\n->scope`);
      log(scopeToString(ast.rootScope));
      log();
    });
  }

  // TODO diff
  // if (argv.diff) printDiff(srcPath, origWgsl, linked);
}

function externalDefines(): Record<string, string> {
  if (!argv.define) return {};
  const pairs = argv.define.map(d => d.toString().split("="));

  const badPair = pairs.find(p => p.length !== 2);
  if (badPair) {
    console.error("invalid define", badPair);
    return {};
  }

  throw new Error("external defines Not implemented");
  const withParsedValues = pairs.map(([k, v]) => [k, parseDefineValue(v)]);
  // return Object.fromEntries(withParsedValues);
}

function parseDefineValue(value: string): string | number | boolean {
  const v = value.toLowerCase();
  if (v === "true") return true;
  if (v === "false") return false;
  if (value === "NaN") return NaN;
  const n = Number.parseFloat(value);
  if (!Number.isNaN(n)) return n;
  return value;
}

function printDiff(modulePath: string, src: string, linked: string): void {
  if (src !== linked) {
    const patch = createTwoFilesPatch(modulePath, "linked", src, linked);
    log(patch);
  } else {
    log(`${modulePath}: linked version matches original source`);
  }
}

function rmBaseDirPrefix(path: string): string {
  const baseDir = argv.baseDir;
  if (baseDir) {
    const found = path.indexOf(baseDir);
    if (found !== -1) {
      return path.slice(found + baseDir.length);
    }
  }
  return path;
}
