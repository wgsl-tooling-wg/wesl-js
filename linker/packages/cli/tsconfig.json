{
  "compilerOptions": {
    "target": "ESNext",
    "lib": ["ESNext"],
    "types": ["node"],
    "module": "ESNext",
    "moduleResolution": "Node",
    "esModuleInterop": false,
    "allowSyntheticDefaultImports": true,
    "skipLibCheck": true,
    "allowJs": false,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "importsNotUsedAsValues": "remove",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "allowImportingTsExtensions": true,
    "noEmit": true,
    "paths": {
      "wesl": ["../linker/src"],
      "mini-parse": ["../mini-parse/src"],
      "mini-parse/test-util": ["../mini-parse/src/test-util"],
      "mini-parse/vitest-util": ["../mini-parse/src/vitest-util"]
    }
  },
  "include": ["src/**/*", "src/viteTypes.d.ts"],
  "exclude": ["node_modules"]
}
