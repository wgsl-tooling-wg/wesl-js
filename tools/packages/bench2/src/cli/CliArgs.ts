import yargs from "yargs";

export type CliArgs = ReturnType<typeof cliArgs>;
/** parse command line arguments for wesl-bench */
export function cliArgs(args: string[]) {
  return yargs(args)
    .option("time", {
      type: "number",
      default: 0.642,
      requiresArg: true,
      describe: "benchmark test duration in seconds",
    })
    .option("cpu", {
      type: "boolean",
      default: false,
      describe: "enable CPU counter measurements (requires root)",
    })
    .option("collect", {
      type: "boolean",
      default: false,
      describe: "force a garbage collection after each test",
    })
    .option("observe-gc", {
      type: "boolean",
      default: true,
      describe: "observe garbage collection via perf_hooks",
    })
    .option("profile", {
      type: "boolean",
      default: false,
      describe: "run once, for attaching a profiler",
    })
    .option("mitata", {
      type: "boolean",
      default: false,
      describe: "run using vanilla mitata profiler",
    })
    .option("tinybench", {
      type: "boolean",
      default: false,
      describe: "run using tinybench library",
    })
    .option("filter", {
      type: "string",
      requiresArg: true,
      describe:
        "run only benchmarks matching this regex or substring (case-insensitive)",
    })
    .option("worker", {
      type: "boolean",
      default: false,
      describe: "run benchmarks in a worker thread for better isolation",
    })
    .help()
    .strict()
    .parseSync();
}
