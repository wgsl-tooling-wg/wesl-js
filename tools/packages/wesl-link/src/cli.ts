import path from "node:path";
import { pathToFileURL } from "node:url";
import { enableTracing, log } from "mini-parse";
import { astToString, link, scopeToString } from "wesl";
import {
  dependencyBundles,
  loadModules,
  versionFromPackageJson,
} from "wesl-tooling";
import yargs from "yargs";
import {
  parsedRegistry,
  parseIntoRegistry,
} from "../../wesl/src/ParsedRegistry.js"; // LATER fix import

type CliArgs = Awaited<ReturnType<typeof parseArgs>>;

export async function cli(rawArgs: string[]): Promise<void> {
  enableTracing(); // so we get more debug info
  const argv = await parseArgs(rawArgs);
  await linkNormally(argv);
}

async function parseArgs(args: string[]) {
  const toolDir = path.join(import.meta.url, "..");
  const appVersion = await versionFromPackageJson(toolDir);
  return yargs(args)
    .version(appVersion)
    .option("src", {
      type: "string",
      default: "./shaders/*.w[eg]sl",
      describe: "WGSL/WESL files to bundle in the package (glob syntax)",
    })
    .option("rootModule", {
      type: "string",
      default: "main",
      describe: "start linking from this module name",
    })
    .option("conditions", {
      type: "array",
      describe: "settings for conditional compilation",
    })
    .option("baseDir", {
      requiresArg: true,
      type: "string",
      default: "./shaders",
      describe: "root directory for shaders",
    })
    .option("projectDir", {
      requiresArg: true,
      type: "string",
      default: ".",
      describe: "directory containing package.json",
    })
    .option("details", {
      type: "boolean",
      default: false,
      hidden: true,
      describe: "show details about parsed files",
    })
    .option("emit", {
      type: "boolean",
      default: true,
      hidden: true,
      describe: "emit linked result",
    })
    .help()
    .parse();
}

async function linkNormally(argv: CliArgs): Promise<void> {
  const { baseDir, projectDir, rootModule: rootModuleName } = argv;
  const weslRoot = baseDir || process.cwd();
  const weslSrc = await loadModules(projectDir, weslRoot, argv.src);
  const projectDirAbs = path.resolve(projectDir, "src.js");
  const projectDirUrl = pathToFileURL(projectDirAbs);
  const libs = await dependencyBundles(weslSrc, projectDirUrl.href);

  const conditionEntries = argv.conditions?.map(c => [c, true]) || [];
  const conditions = Object.fromEntries(conditionEntries);

  if (argv.emit) {
    const linked = await link({ weslSrc, rootModuleName, libs, conditions });
    log(linked.dest);
  }

  if (argv.details) {
    const registry = parsedRegistry();
    try {
      parseIntoRegistry(weslSrc, registry, "package");
    } catch (e) {
      console.error(e);
    }
    Object.entries(registry.modules).forEach(([modulePath, ast]) => {
      log(`---\n${modulePath}`);
      log(`\n->ast`);
      log(astToString(ast.moduleElem));
      log(`\n->scope`);
      log(scopeToString(ast.rootScope));
      log();
    });
  }
}
