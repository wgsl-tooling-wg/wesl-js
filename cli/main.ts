#!/usr/bin/env -S deno run --allow-read --allow-write
import { cli } from "./cli.ts";

if (import.meta.main) {
    await cli(Deno.args);
}
