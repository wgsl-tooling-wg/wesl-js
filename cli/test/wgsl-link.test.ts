import { expect, test } from "vitest";
import { stub } from "@std/testing/mock";
import { cli } from "../cli.ts";
import { assertSnapshot } from "@std/testing/snapshot";

/** so vitest triggers when these files change */
// Deno doesn't support this at the moment :(
// https://github.com/denoland/deno/issues/17994#issuecomment-1879636936 hasn't materialized yet
// import("./src/test/wgsl/main.wgsl?raw");
// import("./src/test/wgsl/util.wgsl?raw");

test("simple link", async (ctx) => {
  using consoleStub = stub(console, "log", () => {});
  await cli(
    `./test/wgsl/main.wgsl 
       ./test/wgsl/util.wgsl`.split(/\s+/),
  );
  const logged = consoleStub.calls[0].args.join("");
  await assertSnapshot(ctx, logged);
});

test.ignore("link with definition", async () => {
  using consoleStub = stub(console, "log", () => {});
  await cli(
    `./test/wgsl/main.wgsl 
       ./test/wgsl/util.wgsl
       --define EXTRA=true`.split(/\s+/),
  );
  const logged = consoleStub.calls[0].args.join("");
  expect(logged).toContain("fn extra()");
});
