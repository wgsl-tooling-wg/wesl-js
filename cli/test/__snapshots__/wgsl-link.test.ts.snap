export const snapshot = {};

snapshot[`simple link 1`] = `
"fn main() {
  foo();
}

fn foo() {
  // fooImpl
}"
`;
